libminifi/include/core/ThreadedRepository.h

```diff
@@ -0,0 +1,86 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <thread>
+
+#include "Repository.h"
+#include "TraceableResource.h"
+
+namespace org::apache::nifi::minifi::core {
+
+class ThreadedRepository : public core::Repository, public core::TraceableResource {
+ public:
+  using Repository::Repository;
+
+  ~ThreadedRepository() override {
+    if (running_.load()) {
+      logger_->log_error("Thread of %s should have been stopped in subclass before ThreadedRepository's destruction", name_);
+    }
+  }
+
+  // Starts repository monitor thread
+  void start() {
+    if (running_.exchange(true)) {
+      return;
+    }
+    if (purge_period_ <= std::chrono::milliseconds(0)) {
+      return;
+    }
+    getThread() = std::thread(&ThreadedRepository::run, this);
```

Comment:
I don't think there is such a restriction, `&ThreadedRepository::run` is a pointer to member function which, together with the `this`, seem to be a suitable arguments for the `std::thread` constructor, the behind the scenes magic (most likely offset into the vtable) makes sure that when called the overridden method based on the objects dynamic type is called

Reply:
You're right. I didn't find anything in the standard that disallows this, and the member function pointer explanation in [this article](https://www.codeproject.com/Articles/7150/Member-Function-Pointers-and-the-Fastest-Possible) suggests that they keep track internally whether it's virtual or non-virtual and single or multiple inheritance.
I'd still replace it with a lambda, simply because that looks simpler to me, but it's OK as is.